README.md
# CXBQN

VM for [BQN](https://mlochbaum.github.io/BQN/).

See [CONTRIBUTING.md](./CONTRIBUTING.md) for information on contributing.

## Building

Requires a compiler capable of C++20.
GCC 9.3.1 is the oldest compiler I have tested it with.

```console
mkdir build
cd build
# Use CXX to choose your compiler
CXX=g++-11 cmake ..
make -j12
./BQN -v
CXBQN 0.11.0 compiled on Dec 25 2021
```

Or if you want to use nix, CXBQN has a flake:
```
$ nix shell github:ashermancinelli/cxbqn
$ BQN -v
CXBQN 0.11.0 compiled on Jan  1 1980
```

## Using

```
$ ./BQN -h
Usage: BQN [options] args

Positional arguments:
args                   	all remaining arguments are passed into the BQN program as •args [default: {}]

Optional arguments:
-h --help              	shows help message and exits [default: false]
-e --execute           	execute a string as BQN code
-v --version           	prints version information and exits [default: false]
-p --execute-and-print 	execute a string as BQN code, pretty print the result
-x --dump-cu           	dump compilation units after compiling [default: false]
-r --repl              	enter REPL [default: false]
-f --file              	execute a string as BQN code
$ ./BQN -e '•Show 5+5'
10
$ ./BQN -p ↕10
⟨ 0 1 2 3 4 5 6 7 8 9 ⟩
$ ./BQN
   foo←5
5
$ echo '•Show 3‿3⥊↕9' > foo.bqn && ./BQN -f foo.bqn
┌─
╵ 0 1 2
  3 4 5
  6 7 8
        ┘
```

See the values of `•listSys` to see what system functions are available in your build of CXBQN.

The REPL also has Readline support if you enable that option at configure time.
When you run cmake, pass the argument `-DCXBQN_READLINE=ON` like so:
```console
mkdir build && cd build
cmake .. -DCXBQN_READLINE=ON
```

If you have a readline library and development headers installed, this will use
them and provide a much nicer REPL environment with variable name completion
like so:

```
$ ./BQN -r
   foo←5
5
   foobar←10
10
   # Tab once for attempted completion
   fo<TAB>
   foo

   # Tab twice for all options if there is no single completion available
   foo<TAB><TAB>
foo     foobar
   foo
```

## Configuring Options

|     CMake Option     |                                                                          Notes                                                                          |
|:--------------------:|:-------------------------------------------------------------------------------------------------------------------------------------------------------:|
| `CXBQN_READLINE`     | Enable Readline support in the BQN repl (recommended)                                                                                                   |
| `CXBQN_MEM_STRATEGY` | Possible values include "shared_ptr", "leak", and "gc". "gc" has not been implemented yet. For best performance, select "leak"                          |
| `CXBQN_BUILD_TESTS`  | Build test suite. Builds take much longer.                                                                                                              |
| `CXBQN_LOG`          | Enable logging. You will have to set the log level above `level::off` to get output. See log file `cxbqn-debug.log` after running BQN or a test runner. |
| `CXBQN_LOGLEVEL`     | Wrapper around spdlog's log levels. `level::debug` will produce an extremely large amount of data (~10-15gib for executing "2+2").                      |
| `CXBQN_DEBUG_VM`     | Output VM debugging information, including the opcode being executed, the program counter, and any opcode arguments.                                    |
| `CXBQN_DEEPCHECKS`   | Enable deeper checks which may produce more helpful errors but will hurt performance.                                                                   |
| `CXBQN_CUDA`         | Enable CUDA execution via `•_CUDAFor`                                                                                                                   |
| `CXBQN_FFI`          | Enable loading dynamic libraries and calling user functions from them.                                                                                  |
| `CXBQN_NATIVE_R0`    | Enable native C++ implementation of some runtime functions from r0.                                                                                     |
| `CXBQN_NATIVE_R1`    | Enable native C++ implementation of some runtime functions from r1.                                                                                     |
| `CXBQN_PLOT`    | Enable matplotlib bindings and the namespace `•cxbqn.plot`. See [`doc/plot.md`](doc/plot.md) for more information.                                                                                     |

## Specification Feature Support

| Spec Item          | Compliance      | Notes                                                                                                                                        |
|--------------------|-----------------|----------------------------------------------------------------------------------------------------------------------------------------------|
| Bytecode           | Full            |                                                                                                                                              |
| Runtime            | BQN-native and r0 | Almost all of r0 has a native implementation. We plan to replace elements of r1 with native functions as well                                                                                 |
| System Functions   | Partial         | See `•listsys` for the supported system values. Good candidate for first contribution.                                                       |
| Namespaces         | Partial         | Using fields of a namespace and destructuring without aliasing work, however mutable namespace fields are not supported. System-provided namespace `•file` has support for `•file.List` and `•file.Lines`.                    |

## Nonstandard Extensions

### GPU Execution

To build with CUDA, enable the CMake option `CXBQN_CUDA` at configure time.
Support for CUDA execution is limited, but it exists!
The primary function for CUDA execution is `•_CUDAFor`, which is a 1-modifier
that performs the primitive `𝕗` on array `𝕩` and optionally `𝕨`.
If you give this functionality a try, feel free to open an issue for any bugs you encounter (there will be many).
Here's an example:

```console
$ cmake .. -DCXBQN_CUDA=ON
$ make -j 12
$ ./BQN -r
   (↕10) × •_CUDAFor ⌽↕10
⟨ 0 8 14 18 20 20 18 14 8 0 ⟩
   10 + •_CUDAFor ↕15
⟨ 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 ⟩
```

[See the WIP design document for parallel execution here](./doc/GPU.md).

<!--
|      Extension     | Completion |                                                                     Notes                                                                    |
|:------------------:|:----------:|:--------------------------------------------------------------------------------------------------------------------------------------------:|
| GPU Execution      | Partial       | Plan to add support for primitive operations on data values (chars, numbers, strings, and arrays of any of these) on GPUs in the near future |
| Threaded Execution | None       | Same plan as above. Needs more discussion                                                                                                    |
| Libraries and Packaging | None       | There has been much discussion on this topic in the Matrix server, but little consensus has been reached. The first library-related feature CXBQN will support will likely look like `strings←•Import "←strings.bqn"` where `strings.bqn` is the strings library from Marshall's [bqn-libs repository](https://github.com/mlochbaum/bqn-libs). This will likely become a *standard library* of sorts, once consensus is reached. The first character of the string argument passed to `•Import` will have some special meaning indicating that the BQN VM ought to look in some special location (eg all directories in the environment variable `$BQNPATH`) or perform some operation before importing the specified library. |
-->

### Plotting

CXBQN supports bindings to matplotlib via [the matplotlib-cpp](https://github.com/lava/matplotlib-cpp) library.
To enable plotting, enable the CMake option `CXBQN_PLOT`.
You must have Python, matplotlib, and numpy installed.

[See `doc/plot.md` for more information.](doc/plot.md)

```console
BQN -e '•cxbqn.plot.Plot ↕10⋄•cxbqn.plot.Show@'
```
![Simple Plotting Example](doc/img/simple_line_graphs.png)

### FFI

[See the documentation for FFI here.](doc/FFI.md)

### Libraries and Packaging

There has been discussion on this topic on the Matrix server, but no consensus has been reached.
CXBQN supports importing from locations defined in the environment variable `BQNPATH`, a `:`-delimited path just like `$PATH`.

The order in which `•Import` searches for files is:

1. Current working directory
1. The `BQNPATH` environment variable (`:` delimited)
1. `<prefix>/share/bqn` where `<prefix>` is the installation prefix of CXBQN
1. `/usr/share/bqn`

<!--
The first character of the string argument passed to `•Import` will have some special meaning indicating that the BQN VM ought to look in some special location (eg all directories in the environment variable `$BQNPATH`) or perform some operation before importing the specified library.
-->

### Custom Namespace `•cxbqn`

Everything under the namespace `•cxbqn` is specific to the CXBQN implementation of BQN.
As of v0.10.4, only the value `•cxbqn.config` is available under this namespace:

```console
$ ./BQN -p '•cxbqn.config'
⟨ ⟨ "CXBQN_READLINE" 1 ⟩ ⟨ "CXBQN_CUDA" 0 ⟩ ⟨ "CXBQN_FFI" 0 ⟩ ⟨ "CXBQN_PLOT" 0 ⟩ ⟩
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning v2](https://semver.org/spec/v2.0.0.html).

## [0.11.0] 12-29-2021

* Add typed arrays
    * This is a very significant development effort and will be a work in progress for a long time, but this puts the infrastructure in place to maintain homogenous arrays which can be worked on directly, which should have a very significant impact on performance when it's in full use.
* formatting of namespaces and other types •Fmt doesn't know about
* plotting utilities under namespace •cxbqn.plot
* update current documentation, add plotting document

## [0.10.4] 12-27-2021

* add native `•file.List`
* add native namespace `•cxbqn` with only `•cxbqn.config`

## [0.10.3] 12-27-2021

* native namespace support for `•file.Lines`

## [0.10.2] 12-24-2021

* cxbqn can now drive ccx.bqn, hopefully making way for full compilation
* Import now searches `BQNPATH`, `<prefix>/share/bqn`, and `/usr/share/bqn`

## [0.10.1] 12-23-2021

* Extreme perf now realized after fixing bug in enabling native runtime

## [0.10.0] 12-22-2021

* Extreme perf improvements from native modifiers in the runtime
* The following runtime values have _some_ native support:
    * NE (≠)
    * Rtack (⊢)
    * Ltack (⊣)
    * JoinTo (∾)
    * EnlistPair (⋈)
    * Take (↑)
    * Drop (↓)
    * Constant (˙)
    * SelfSwap (˜)
    * Each (¨)
    * Fold (´)
    * Atop (∘)
    * Over (○)
    * Choose (◶)
    * Before (⊸)
    * After (⟜)
    * Repeat (⍟)

## [0.9.1] 12-22-2021

* Begin work on native runtime
* Add some simple native runtime functions

## [0.9.0] 12-22-2021

* Add the cxbqn-config utility and install all targets
* Update documentation
* Add FFI as `•_FFI_`

## [0.8.2] 12-21-2021

* Add swath of CUDA functions under `•_CUDAFor`

## [0.8.1] 11-29-2021

* Use observer ptrs for most pointers, leak just like CBQN. Huge perf boost.

## [0.8.0] 11-28-2021

* Add memory mangement policies

## [0.7.0] 11-24-2021

* Significant improvements to the driver
* Working repl
* Readline support for the repl
* made public!

## [0.6.0] 11-19-2021

* Significant improvements to the driver
* Significant performance improvements

## [0.5.0] 11-18-2021

* Significant improvements to the driver
* support for gcc 9

## [0.4.1] 11-18-2021

* Formatter working
* Compiler compiling string
* Passing under, undo, and identity tests

## [0.4.0] 11-18-2021

* Pretty backtraces
* passing l1, 2, 3, 4, 5, 6 prim tests
* building compiler
* added support for source locations

## [0.3.2] 11-14-2021

* Passing l0 prim tests, still failing a few later prim tests.

## [0.3.0] 11-9-2021

* Fixed significant memory management issues. Now fit runtime in 2-4mib of ram.

## [0.2.0] 11-11-2021

* passing all simple and primitive tests
* successfully loading runtime
* added some memory management
    * memory still going wild when loading the compiler

## [0.1.0] 11-4-2021

* VM passing 100% of bytecode tests as described in `BQN/test/cases/bytecode.bqn`
* VM passing many simple and primitive tests.

## [0.0.2] 10-24-2021

* `Type` primitive
* infrastructure for dynamically generating bytecode tests from upstream BQN repo

## [0.0.1] 10-22-2021

* Initial repo. Referring to [ebqn-rs](https://github.com/cannadayr/ebqn-rs/blob/master/crs.bqn) for structures and [the old bqn javascript runtime](https://github.com/mlochbaum/BQN/blob/f74d9223ef880f2914030c2375f680dcc7e8c92b/bqn.js#L23) to get started.
* Formatting targets
* README, LICENSE, CHANGELOG


src/bi.cpp
#include <cmath>
#include <cxbqn/comp_utils.hpp>
#include <cxbqn/cxbqn.hpp>
#include <cxbqn/debug.hpp>

namespace cxbqn::provides {
O<types::Array> get_provides() {
  CXBQN_DEBUG("provides::get_provides");
  auto prov = make_shared<Array>(23);

  prov->values[0] = bi_Type();
  prov->values[1] = bi_Fill();
  prov->values[2] = bi_Log();
  prov->values[3] = bi_GroupLen();
  prov->values[4] = bi_GroupOrd();
  prov->values[5] = bi_Assert();
  prov->values[6] = bi_Plus();
  prov->values[7] = bi_Minus();
  prov->values[8] = bi_Mul();
  prov->values[9] = bi_Div();
  prov->values[10] = bi_Power();
  prov->values[11] = bi_Floor();
  prov->values[12] = bi_EQ();
  prov->values[13] = bi_LE();
  prov->values[14] = bi_FNE();
  prov->values[15] = bi_Deshape();
  prov->values[16] = bi_Pick();
  prov->values[17] = bi_Range();
  prov->values[18] = bi_Table();
  prov->values[19] = bi_Scan();
  prov->values[20] = bi_FillBy();
  prov->values[21] = bi_Valence();
  prov->values[22] = bi_Catch();

  return prov;
}

static O<Array> rt = nullptr;
O<Array> get_runtime_cached() {
  if (nullptr != rt)
    return rt;
  rt = get_runtime();
  return get_runtime_cached();
}

O<Array> get_runtime() {
  CXBQN_DEBUG("provides::get_runtime");

  const auto provide = provides::get_provides()->values;
  static CompileParams p(
#include <cxbqn/__/compiled_runtime>
  );
  auto ret = vm::run(p.bc, p.consts.to_arr(), p.blk_defs, p.bodies);
  auto runtime_ret = std::dynamic_pointer_cast<Array>(ret.v);

  return std::dynamic_pointer_cast<Array>(runtime_ret->values[0]);
}

static O<Array> rtsp = nullptr;
O<Array> get_runtime_setprims_cached() {
  if (nullptr != rtsp)
    return rtsp;
  rtsp = get_runtime_setprims();
  return get_runtime_setprims_cached();
}

O<Array> get_runtime_setprims() {
  CXBQN_DEBUG("provides::get_runtime_setprims");

  // First we run the runtime to get the original output
  const auto provide = provides::get_provides()->values;
  static CompileParams p(
#include <cxbqn/__/compiled_runtime>
  );
  auto ret = vm::run(p.bc, p.consts.to_arr(), p.blk_defs, p.bodies);

  // Decompose the result to get the array with just the runtime
  auto runtime_ret = std::dynamic_pointer_cast<Array>(ret.v);
  auto runtime_raw = std::dynamic_pointer_cast<Array>(runtime_ret->values[0]);

  auto setprims = runtime_ret->values[1];

  // Inform the two latter builtins of the runtime so they can refer to it
  auto decompose = make_shared<Decompose>(runtime_raw);
  auto primind = make_shared<PrimInd>(runtime_raw);

  setprims->call(1, {setprims, O<Array>(new Array({decompose, primind})), bi_Nothing()});

  return runtime_raw;
}

O<Array> get_runtime_bionly() {
  CXBQN_DEBUG("provides::get_runtime_bionly");
  auto rt = make_shared<Array>(60);

  rt->values[0] = bi_Plus();
  rt->values[1] = bi_Minus();
  rt->values[2] = bi_Mul();
  rt->values[3] = bi_Div();
  rt->values[4] = bi_Power();
  rt->values[5] = bi_Root();
  rt->values[6] = bi_Floor();
  rt->values[7] = bi_Ceil();
  rt->values[8] = bi_Stile();
  rt->values[9] = bi_Not();
  rt->values[10] = bi_And();
  rt->values[11] = bi_Or();
  rt->values[12] = bi_LT();
  rt->values[13] = bi_GT();
  rt->values[14] = bi_NE();
  rt->values[15] = bi_EQ();
  rt->values[16] = bi_LE();
  rt->values[17] = bi_GE();
  rt->values[18] = bi_FEQ();
  rt->values[19] = bi_FNE();
  rt->values[20] = bi_Ltack();
  rt->values[21] = bi_Rtack();
  rt->values[22] = bi_Deshape();
  rt->values[27] = bi_Range();
  rt->values[36] = bi_Pick();
  rt->values[47] = bi_Table();
  rt->values[51] = bi_Scan();

  return rt;
}

#define BI_DEF(T)                                                              \
  O<Value> bi_##T() {                                                          \
    static T v;                                                                \
    return std::shared_ptr<T>(&v, [](T *) {});                                 \
  }
BI_DEF(Plus);
BI_DEF(Minus);
BI_DEF(Mul);
BI_DEF(Div);
BI_DEF(Power);
BI_DEF(Root);
BI_DEF(Floor);
BI_DEF(Ceil);
BI_DEF(Stile);
BI_DEF(Not);
BI_DEF(And);
BI_DEF(Or);
BI_DEF(LT);
BI_DEF(GT);
BI_DEF(NE);
BI_DEF(EQ);
BI_DEF(LE);
BI_DEF(GE);
BI_DEF(FEQ);
BI_DEF(FNE);
BI_DEF(Ltack);
BI_DEF(Rtack);
BI_DEF(ArrayDepth);
BI_DEF(Type);
BI_DEF(Table);
BI_DEF(Fill);
BI_DEF(Log);
BI_DEF(Assert);
BI_DEF(Range);
BI_DEF(Pick);
BI_DEF(Shape);
BI_DEF(Deshape);
BI_DEF(Scan);
BI_DEF(GroupLen);
BI_DEF(GroupOrd);
BI_DEF(FillBy);
BI_DEF(Valence);
BI_DEF(Catch);
// BI_DEF(Decompose);
// BI_DEF(PrimInd);

} // namespace cxbqn::provides
